File: android/app/src/main/java/com/vxin/app/core/storage/OutboxStore.kt
package com.vxin.app.core.storage

import android.content.Context
import android.content.SharedPreferences
import com.vxin.app.data.model.LocalMsgStatus
import com.vxin.app.data.model.Message
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import javax.inject.Inject
import javax.inject.Singleton

/** Only the current session can persist/read its own pending messages. */
@Singleton
class OutboxStore internal constructor(
    private val prefs: SharedPreferences,
    private val scopes: MessageScopeProvider,
) {
    @Inject constructor(@ApplicationContext context: Context, scopes: MessageScopeProvider) :
        this(context.getSharedPreferences("vxin_outbox", Context.MODE_PRIVATE), scopes)
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val listSerializer = ListSerializer(Message.serializer())

    fun load(conversationId: String, scope: MessageScope? = scopes.current()): List<Message> =
        loadRaw(conversationId, scope).map { it.copy(localStatus = LocalMsgStatus.FAILED, clientMsgId = it.id) }

    fun upsert(conversationId: String, msg: Message, scope: MessageScope? = scopes.current()) {
        if (!scopes.isCurrent(scope) || msg.type != "text" ||
            scope?.owns(msg.sender_id, msg.conversation_id, conversationId) != true) return
        val list = loadRaw(conversationId, scope).toMutableList()
        val idx = list.indexOfFirst { it.id == msg.id }
        if (idx >= 0) list[idx] = msg else list.add(msg)
        save(conversationId, list.takeLast(50), scope)
    }

    fun remove(conversationId: String, msgId: String, scope: MessageScope? = scopes.current()) {
        val list = loadRaw(conversationId, scope)
        val next = list.filterNot { it.id == msgId }
        if (next.size != list.size) save(conversationId, next, scope)
    }

    private fun loadRaw(conversationId: String, scope: MessageScope?): List<Message> {
        if (conversationId.isBlank() || !scopes.isCurrent(scope)) return emptyList()
        val raw = prefs.getString("${scope!!.key}:$conversationId", null) ?: return emptyList()
        return runCatching { json.decodeFromString(listSerializer, raw) }.getOrDefault(emptyList())
            .filter { it.type == "text" && scope.owns(it.sender_id, it.conversation_id, conversationId) }
    }

    private fun save(conversationId: String, list: List<Message>, scope: MessageScope?) {
        if (!scopes.isCurrent(scope)) return
        val key = "${scope!!.key}:$conversationId"
        prefs.edit().apply {
            if (list.isEmpty()) remove(key) else putString(key, json.encodeToString(listSerializer, list))
        }.apply()
    }

    fun clearScope(scope: MessageScope? = scopes.current()) {
        val prefix = scope?.let { "${it.key}:" }
        prefs.edit().apply {
            prefs.all.keys.filter { !it.startsWith("v2_") || (prefix != null && it.startsWith(prefix)) }
                .forEach { remove(it) }
        }.apply()
    }
}

File: ios/Vxin/Core/Storage/OutboxStore.swift
import Foundation

/// 失败消息「待发件箱」——让发送失败的文本消息在切走会话 / 杀进程重启后依然不丢失，
/// 对齐 Web / Android 的 outbox 体验。按 conversationId 持久化到 UserDefaults。
/// 只存纯文本（type=="text"）；每会话最多 50 条防膨胀。
final class OutboxStore {
    static let shared = OutboxStore()
    private let defaults: UserDefaults
    private let currentScope: () -> MessageScope?
    init(defaults: UserDefaults = .standard, currentScope: @escaping () -> MessageScope? = { MessageScope.current }) {
        self.defaults = defaults
        self.currentScope = currentScope
    }

    private let prefix = "vxin_outbox_"
    private let maxPerConv = 50

    /// 可持久化的失败消息快照（Message 仅 Decodable，这里用独立 Codable 结构）
    private struct Item: Codable {
        let id: String
        let conversationId: String
        let senderId: String
        let content: String
        let replyToId: String?
        let createdAt: Double
        let replyToSenderName: String?
        let replyToId2: String?
        let replyToType: String?
        let replyToContent: String?
    }

    // MARK: - Public

    /// 读取某会话的待发件箱，还原为 failed 态的 Message 列表
    func load(_ conversationId: String, scope: MessageScope? = MessageScope.current) -> [Message] {
        loadItems(conversationId, scope: scope).map { item in
            let reply: ReplyPreview? = item.replyToId2.map {
                ReplyPreview(id: $0, type: item.replyToType ?? "text",
                             content: item.replyToContent ?? "", senderName: item.replyToSenderName ?? "")
            }
            var m = Message(optimisticText: item.id, conversationId: item.conversationId,
                            senderId: item.senderId, content: item.content,
                            replyToId: item.replyToId, replyTo: reply, clientMsgId: item.id)
            m.localStatus = LocalMsgStatus.failed
            m.createdAt = item.createdAt
            return m
        }
    }

    /// 新增/更新一条失败消息（按 id 去重；仅文本）
    func upsert(_ conversationId: String, _ msg: Message, scope: MessageScope? = MessageScope.current) {
        guard let scope, scope == currentScope(), msg.type == "text",
              scope.owns(senderId: msg.senderId, messageConversationId: msg.conversationId, conversationId: conversationId) else { return }
        var items = loadItems(conversationId, scope: scope)
        let item = Item(
            id: msg.id, conversationId: conversationId, senderId: msg.senderId,
            content: msg.content, replyToId: msg.replyToId, createdAt: msg.createdAt,
            replyToSenderName: msg.replyTo?.senderName, replyToId2: msg.replyTo?.id,
            replyToType: msg.replyTo?.type, replyToContent: msg.replyTo?.content
        )
        if let idx = items.firstIndex(where: { $0.id == msg.id }) { items[idx] = item }
        else { items.append(item) }
        save(conversationId, Array(items.suffix(maxPerConv)), scope: scope)
    }

    /// 消息成功送达后移除（按 id）
    func remove(_ conversationId: String, _ msgId: String, scope: MessageScope? = MessageScope.current) {
        guard !conversationId.isEmpty else { return }
        let items = loadItems(conversationId, scope: scope)
        let next = items.filter { $0.id != msgId }
        if next.count != items.count { save(conversationId, next, scope: scope) }
    }

    // MARK: - Private

    func clearScope(_ scope: MessageScope? = MessageScope.current) {
        for key in defaults.dictionaryRepresentation().keys where key.hasPrefix(prefix) {
            if !key.hasPrefix(prefix + "v2_") || scope.map({ key.hasPrefix(prefix + $0.key + ":") }) == true {
                defaults.removeObject(forKey: key)
            }
        }
    }

    private func loadItems(_ conversationId: String, scope: MessageScope?) -> [Item] {
        guard let scope, scope == currentScope(), !conversationId.isEmpty else { return [] }
        guard let data = defaults.data(forKey: prefix + scope.key + ":" + conversationId) else { return [] }
        return ((try? JSONDecoder().decode([Item].self, from: data)) ?? []).filter {
            scope.owns(senderId: $0.senderId, messageConversationId: $0.conversationId, conversationId: conversationId)
        }
    }

    private func save(_ conversationId: String, _ items: [Item], scope: MessageScope?) {
        guard let scope, scope == currentScope() else { return }
        let key = prefix + scope.key + ":" + conversationId
        if items.isEmpty { defaults.removeObject(forKey: key); return }
        if let data = try? JSONEncoder().encode(items) { defaults.set(data, forKey: key) }
    }
}
